### PR TITLE
feat: Pending offsets handler

### PR DESCRIPTION
**Pull request by dougolima**
# Description

Adds a handler to be executed when the consumer has pending offsets to be committed

## How Has This Been Tested?

Unit tests were created

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my own code
-   [x] I have added tests to cover my changes
-   [ ] I have made corresponding changes to the documentation

### Disclaimer

By sending us your contributions, you are agreeing that your contribution is made subject to the terms of our [Contributor Ownership Statement](https://github.com/Farfetch/.github/blob/master/COS.md)

### CODE REVIEW

File: src/KafkaFlow/Consumers/PartitionOffsets.cs [hunk 1]
```diff
@@ -7,42 +7,42 @@ namespace KafkaFlow.Consumers
...
-            lock (this.offsetsOrder)
+            lock (this.receivedOffsets)
             {
-                this.offsetsOrder.AddLast(offset);
+                this.receivedOffsets.AddLast(offset);
             }
         }
 
-        public bool ShouldUpdateOffset(long newOffset)
+        public bool ShouldCommit(long offset)
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/PartitionOffsets.cs:24]: I think ShouldSendToCommitter() better

File: src/KafkaFlow/Consumers/OffsetManager.cs [hunk 2]
```diff
@@ -20,7 +20,7 @@ internal class OffsetManager : IOffsetManager, IDisposable
                 partition => new PartitionOffsets());
         }
 
-        public void StoreOffset(TopicPartitionOffset offset)
+        public void Commit(TopicPartitionOffset offset)
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/OffsetManager.cs:23]: I would change this name because we are not committing here. Maybe EnqueueToCommit()

File: src/KafkaFlow/Consumers/ConsumerWorkerPool.cs [hunk 3]
```diff
@@ -22,22 +25,24 @@ internal class ConsumerWorkerPool : IConsumerWorkerPool
...
-            IMiddlewareExecutor middlewareExecutor,
-            ILogHandler logHandler,
-            Factory<IDistributionStrategy> distributionStrategyFactory)
+            IConsumerConfiguration consumerConfiguration,
+            ILogHandler logHandler)
         {
             this.consumer = consumer;
             this.dependencyResolver = dependencyResolver;
-            this.middlewareExecutor = middlewareExecutor;
+            this.middlewareExecutor = new MiddlewareExecutor(consumerConfiguration.MiddlewaresConfigurations);
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/ConsumerWorkerPool.cs:33]: Why this change?

File: src/KafkaFlow/Consumers/OffsetCommitter.cs [hunk 4]
```diff
@@ -13,11 +13,14 @@ internal class OffsetCommitter : IOffsetCommitter
         private readonly ILogHandler logHandler;
 
         private readonly Timer commitTimer;
+        private readonly IReadOnlyList<Timer> statisticsTimers;
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/OffsetCommitter.cs:16]: I would create just one handler to this statistic, many Timers can lead to performance degradation

File: src/KafkaFlow/Consumers/OffsetCommitter.cs [hunk 5]
```diff
@@ -80,7 +106,7 @@ private void CommitHandler()
             {
                 this.logHandler.Warning(
                     "Error Commiting Offsets",
-                    new { ErrorMessage = e.Message });
+                    new {ErrorMessage = e.Message});
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/OffsetCommitter.cs:109]: I think you have to configure your IDE preferences to include these spaces

File: src/KafkaFlow/Consumers/PartitionOffsets.cs [hunk 6]
```diff
@@ -7,42 +7,42 @@ namespace KafkaFlow.Consumers
...
-        private readonly LinkedList<long> offsetsOrder = new();
+        private readonly SortedSet<long> processedOffsets = new();
+        private readonly LinkedList<long> receivedOffsets = new();
 
-        public long LastOffset { get; private set; } = -1;
+        public long LastCommittedOffset { get; private set; } = -1;
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public void AddOffset(long offset)
+        public void Store(long offset)
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/PartitionOffsets.cs:16]: I think Store() can lead to confusion here because "store" is the term used to store an offset to be committed

File: src/KafkaFlow/Consumers/IOffsetCommitter.cs [hunk 7]
```diff
@@ -5,6 +5,6 @@ namespace KafkaFlow.Consumers
 
     internal interface IOffsetCommitter : IDisposable
     {
-        void StoreOffset(TopicPartitionOffset tpo);
+        void Commit(TopicPartitionOffset tpo);
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/IOffsetCommitter.cs:8]: I would change this name because we are not committing here. Maybe EnqueueToCommit()

File: src/KafkaFlow/Consumers/PartitionOffsets.cs [hunk 8]
```diff
@@ -7,42 +7,42 @@ namespace KafkaFlow.Consumers
 
     internal class PartitionOffsets
     {
-        private readonly SortedSet<long> pendingOffsets = new();
-        private readonly LinkedList<long> offsetsOrder = new();
+        private readonly SortedSet<long> processedOffsets = new();
+        private readonly LinkedList<long> receivedOffsets = new();
 
-        public long LastOffset { get; private set; } = -1;
+        public long LastCommittedOffset { get; private set; } = -1;
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/PartitionOffsets.cs:13]: Maybe OffsetToCommit?

**Review by filipeesch — CHANGES_REQUESTED** [review]

**Comment by filipeesch** [comment]
change the commit message to "feat: adds Pending offsets statistics handler"

File: samples/KafkaFlow.Sample/Program.cs [hunk 9]
```diff
@@ -50,6 +50,22 @@ private static async Task Main()
...
+                                                    {
+                                                        Offsets = offsets.Select(
+                                                            o => new
+                                                            {
+                                                                Partition = o.Value.Partition.Value,
+                                                                Offset = o.Value.Offset.Value,
+                                                                Topic = o.Value.Partition
+                                                            })
+                                                    }),
+                                        1000)
```
Comment (filipeesch) [inline samples/KafkaFlow.Sample/Program.cs:68]: bad indentation here

File: src/KafkaFlow/Extensions/ConfigurationBuilderExtensions.cs [hunk 10]
```diff
@@ -82,6 +82,22 @@ public static IConsumerConfigurationBuilder WithConsumerConfig(this IConsumerCon
...
+        /// <param name="intervalMs">The execution interval in milliseconds</param>
+        /// <returns></returns>
+        public static IConsumerConfigurationBuilder OnPendingOffsets(
+            this IConsumerConfigurationBuilder builder,
+            Action<IDependencyResolver, IReadOnlyDictionary<(string, int), TopicPartitionOffset>> pendingOffsetsHandler,
+            int intervalMs)
+        {
+            ((ConsumerConfigurationBuilder)builder).OnPendingOffsets(pendingOffsetsHandler, intervalMs);
+            return builder;
+        }
```
Comment (filipeesch) [inline src/KafkaFlow/Extensions/ConfigurationBuilderExtensions.cs:99]: I think WithPendingOffsetsStatisticsHandler will be better

File: src/KafkaFlow/Extensions/ConfigurationBuilderExtensions.cs [hunk 11]
```diff
@@ -82,6 +82,22 @@ public static IConsumerConfigurationBuilder WithConsumerConfig(this IConsumerCon
...
         }
 
+        /// <summary>
+        /// Adds a handler to be executed when KafkaFlow has pending offsets to be committed
+        /// </summary>
+        /// <param name="builder">The configuration builder</param>
+        /// <param name="pendingOffsetsHandler">A handler for the consumer pending offsets state</param>
+        /// <param name="intervalMs">The execution interval in milliseconds</param>
+        /// <returns></returns>
+        public static IConsumerConfigurationBuilder OnPendingOffsets(
```
Comment (filipeesch) [inline src/KafkaFlow/Extensions/ConfigurationBuilderExtensions.cs:92]: Why an extension method instead of a common method?

**Review by filipeesch — CHANGES_REQUESTED** [review]

File: src/KafkaFlow/Extensions/ConfigurationBuilderExtensions.cs (hunk 11 above)
Comment (dougolima) [reply src/KafkaFlow/Extensions/ConfigurationBuilderExtensions.cs:92]: it couldn't be in the abstraction layer because of Confluent.Kafka.TopicPartitionOffset Action parameter

**Review by filipeesch — APPROVED** [review]

**Review by filipeesch — APPROVED** [review]
